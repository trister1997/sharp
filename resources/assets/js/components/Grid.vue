<template>
    <div class="SharpGrid">
        <div v-for="row in rows" class="SharpGrid__row row">
            <div v-for="col in row" :class="colClasses(col)" class="SharpGrid__col" v-empty-class="'SharpGrid__col--empty'">
                <slot v-bind="col" />
            </div>
        </div>
    </div>
</template>

<script>
    import emptyClass from '../directives/EmptyClass';

    export default {
        name: 'SharpGrid',

        props: {
            rows: { // 2D array [row][col]
                type: Array,
                required: true
            }
        },
        methods: {
            colClasses({ size, sizeXS }) {
                const hasSize = !!size;
                return {
                    [`col-${sizeXS}`]: sizeXS,
                    [`col-md-${size}`]: hasSize,
                    'col-md': !hasSize
                }
            }
        },
        directives: {
            'empty-class': emptyClass
        }
    }
</script>
