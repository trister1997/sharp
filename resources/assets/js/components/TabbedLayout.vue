<template>
    <div class="SharpTabbedLayout">
        <template v-if="showTabs">
            <SharpTabs>
                <template slot="nav-prepend"><slot name="nav-prepend"></slot></template>
                <SharpTab v-for="(tab,i) in layout.tabs" :title="tab.title" :key="i">
                    <slot v-bind="tab"></slot>
                </SharpTab>
            </SharpTabs>
        </template>
        <template v-else>
            <div><slot name="nav-prepend"></slot></div>
            <div v-for="tab in layout.tabs">
                <slot v-bind="tab"></slot>
            </div>
        </template>
    </div>
</template>

<script>
    import SharpTabs from './Tabs';
    import SharpTab from './Tab';

    export default {
        name:'SharpTabbedLayout',
        props : {
            layout: Object,
        },
        provide() {
            if(!this.showTabs) {
                return { $tab: false }
            }
        },
        components: {
            SharpTabs,
            SharpTab,
        },
        computed: {
            showTabs() {
                return this.layout.tabbed && this.layout.tabs.length>1;
            }
        }
    }
</script>
