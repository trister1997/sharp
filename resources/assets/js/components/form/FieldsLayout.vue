<template>
    <sharp-grid :rows="layout">
        <template slot-scope="fieldLayout">
            <slot v-if="!fieldLayout.legend" v-bind="fieldLayout"></slot>

            <fieldset class="SharpForm__fieldset" v-else v-show="isFieldsetVisible(fieldLayout)">
                <div class="SharpModule__inner">
                    <div class="SharpModule__header">
                        <legend class="SharpModule__title">{{fieldLayout.legend}}</legend>
                    </div>
                    <div class="SharpModule__content">
                        <sharp-fields-layout :layout="fieldLayout.fields">
                            <template slot-scope="fieldset">
                                <slot v-bind="fieldset"></slot>
                            </template>
                        </sharp-fields-layout>
                    </div>
                </div>
            </fieldset>
        </template>
    </sharp-grid>
</template>

<script>
    import SharpGrid from '../Grid';

    export default {
        name:'SharpFieldsLayout',

        components: {
            SharpGrid
        },

        props: {
            layout: { // 2D array fields [ligne][col]
                type: Array,
                required: true
            },
            visible : {
                type: Object,
                default: () => ({})
            }
        },

        data() {
            return {
                fieldsetMap: {}
            }
        },

        methods: {
            isFieldsetVisible(fieldLayout) {
                let { id, fields } = fieldLayout;

                let map = this.fieldsetMap[id] || (this.fieldsetMap[id] = [].concat.apply([],fields));
                return map.some(f => this.visible[f.key]);
            }
        }
    }
</script>