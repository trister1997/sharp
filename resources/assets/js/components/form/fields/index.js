import SharpAutocomplete from './Autocomplete.vue';
import SharpTextarea from './Textarea.vue';
import SharpText from './Text.vue';
import SharpMarkdown from './markdown/Markdown.vue';
import SharpNumber from './Number.vue';
import SharpUpload from './upload/Upload.vue';
import SharpTagInput from './Tags.vue';
import SharpDate from './date/Date.vue';
import SharpCheck from './Check.vue';
import SharpList from './list/List.vue';
import SharpSelect from './Select.vue';
import SharpHtml from './Html.vue';
import SharpGeolocation from './geolocation/Geolocation';
import SharpTrix from './wysiwyg/TrixEditor.vue';
import SharpDateRange from './date-range/DateRange';

export default {
    'autocomplete' : SharpAutocomplete,
    'text'         : SharpText,
    'textarea'     : SharpTextarea,
    'markdown'     : SharpMarkdown,
    'number'       : SharpNumber,
    'upload'       : SharpUpload,
    'tags'         : SharpTagInput,
    'date'         : SharpDate,
    'check'        : SharpCheck,
    'list'         : SharpList,
    'select'       : SharpSelect,
    'html'         : SharpHtml,
    'geolocation'  : SharpGeolocation,
    'wysiwyg'      : SharpTrix,
    'daterange'    : SharpDateRange,
};