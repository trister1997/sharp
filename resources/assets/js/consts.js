export let BASE_URL = `${BASE_PATH}`;
export let API_PATH = `/${BASE_URL}/api`;
export let UPLOAD_URL = `${API_PATH}/upload`;