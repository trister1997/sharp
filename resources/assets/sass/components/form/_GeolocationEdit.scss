.SharpGeolocationEdit {
  &--loading {
    .SharpGeolocationEdit__input {
      padding-right: 2rem;
    }
  }
  .SharpGeolocationEdit__loading {
    position: absolute;
    top:0;
    right:0;
    bottom:0;
    margin: auto;
  }
  .leaflet-container {
    cursor: crosshair;
  }

  &__map {
    padding-bottom: 80%;
    max-width: 100%;
  }
}