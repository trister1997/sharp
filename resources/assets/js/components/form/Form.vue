<template>
    <div class="SharpForm">
        <template v-if="ready">
            <div v-show="hasErrors" class="SharpNotification SharpNotification--error" role="alert">
                <div class="SharpNotification__details">
                    <div class="SharpNotification__text-wrapper">
                        <p class="SharpNotification__title">{{ l('form.validation_error.title') }}</p>
                        <p class="SharpNotification__subtitle">{{ l('form.validation_error.description') }}</p>
                    </div>
                </div>
            </div>
            <sharp-tabbed-layout :layout="layout" ref="tabbedLayout">
                <!-- Tab -->
                <template slot-scope="tab">
                    <sharp-grid :rows="[tab.columns]" ref="columnsGrid">
                        <!-- column -->
                        <template slot-scope="column">
                            <sharp-fields-layout v-if="fields" :layout="column.fields" :visible="fieldVisible" ref="fieldLayout">
                                <!-- field -->
                                <template slot-scope="fieldLayout">
                                    <sharp-field-display :field-key="fieldLayout.key"
                                                         :context-fields="isReadOnly ? readOnlyFields : fields"
                                                         :context-data="data"
                                                         :field-layout="fieldLayout"
                                                         :locale="locale"
                                                         :error-identifier="fieldLayout.key"
                                                         :config-identifier="fieldLayout.key"
                                                         :update-data="updateData"
                                                         :update-visibility="updateVisibility"
                                                         ref="field">
                                    </sharp-field-display>
                                </template>
                            </sharp-fields-layout>
                        </template>
                    </sharp-grid>
                </template>
            </sharp-tabbed-layout>
        </template>
    </div>
</template>

<script>
    import * as util from '../../util';
    import { API_PATH } from '../../consts';

    import { ActionEvents, ReadOnlyFields, Localization } from '../../mixins';

    import DynamicView from '../DynamicViewMixin';

    import TabbedLayout from '../TabbedLayout'
    import Grid from '../Grid';
    import FieldsLayout from './FieldsLayout.vue';


    const noop = ()=>{}

    export default {
        name:'SharpForm',
        extends: DynamicView,

        mixins: [ActionEvents, ReadOnlyFields('fields'), Localization],

        components: {
            [TabbedLayout.name]: TabbedLayout,
            [FieldsLayout.name]: FieldsLayout,
            [Grid.name]: Grid,
        },


        props:{
            entityKey: String,
            instanceId: String,

            /// Extras props for customization
            independant: {
                type:Boolean,
                default: false
            },
            ignoreAuthorizations: Boolean,
            props: Object
        },

        inject:['actionsBus'],

        provide() {
            return {
                $form:this
            }
        },

        data() {
            return {
                fields: null,
                config: null,
                authorizations: null,
                redirect: true,
                deletable: true,

                errors:{},
                locale: '',

                fieldVisible: {},
                curFieldsetId:0,

                pendingJobs: []
            }
        },
        computed: {
            apiPath() {
                let path = `${API_PATH}/form/${this.entityKey}`;
                if(this.instanceId) path+=`/${this.instanceId}`;
                return path;
            },
            localized() {
                return this.config && Array.isArray(this.config.locales);
            },
            isCreation() {
                return !this.instanceId;
            },
            isReadOnly() {
                return !this.ignoreAuthorizations && !(this.isCreation ? this.authorizations.create : this.authorizations.update);
            },
            // don't show loading on creation
            synchronous() {
                return this.independant;
            },
            hasErrors() {
                return Object.keys(this.errors).some(errorKey => !this.errors[errorKey].cleared);
            },

            baseEntityKey() {
                return this.entityKey.split(':')[0];
            },

            downloadLinkBase() {
                return `${API_PATH}/download/${this.entityKey}/${this.instanceId}`;
            }
        },
        methods: {
            fieldErrors(key) {
                if(this.fields[key].localized) {
                    return (this.errors[key]||{})[this.locale];
                }
                return this.errors[key];
            },
            updateData(key,value) {
                if(this.fields[key].localized) {
                    this.$set(this.data[key],this.locale,value);
                }
                else this.$set(this.data,key,value);
            },
            updateVisibility(key, visibility) {
                this.$set(this.fieldVisible, key, visibility);
            },
            mount({fields, layout, data={}, config={}, authorizations={}, redirect = true, deletable = true}) {
                this.fields = fields;
                this.layout = this.patchLayout(layout);
                this.data = data;
                this.config = config;
                this.authorizations = authorizations;
                this.redirect = redirect;
                this.deletable = deletable;


                this.fieldVisible = Object.keys(this.fields).reduce((res, fKey) => {
                    res[fKey] = true;
                    return res;
                },{})
            },
            handleError({response}) {
                if(response.status===422)
                    this.errors = response.data.errors || {};
            },

            patchLayout(layout) {
                let curFieldsetId = 0;
                let mapFields = layout => {
                    if(layout.legend)
                        layout.id = `${curFieldsetId++}#${layout.legend}`;
                    else if(layout.fields)
                        layout.fields.forEach(row => {
                            row.forEach(mapFields);
                        });
                };
                layout.tabs.forEach(t => t.columns.forEach(mapFields));
                return layout;
            },

            init() {
                if(this.independant) {
                    this.mount(this.props);
                    this.ready = true;
                }
                else {
                    if(this.entityKey) {
                        this.get().then(_=>this.setupActionBar());
                    }
                    else util.error('no entity key provided');
                }
            },

            setupActionBar({ disable=false ,setLocale=true }={}) {
                const showSubmitButton = this.isCreation ? this.authorizations.create : this.authorizations.update;

                this.actionsBus.$emit('setup', {
                    locales: null, //this.config.locales,
                    showSubmitButton: showSubmitButton && !disable,
                    showDeleteButton: !this.isCreation && this.authorizations.delete && !disable && this.deletable,
                    showBackButton: this.isReadOnly,
                    opType: this.isCreation ? 'create' : 'update'
                });

                if(setLocale && this.config.locales) {
                    this.actionsBus.$emit('localeChanged', this.config.locales[0]);
                }
            },
            redirectToList({ restoreContext=true }={}) {
                if (this.redirect) {
                    location.href = `/sharp/list/${this.baseEntityKey}${restoreContext ? '?restore-context=1' : ''}`
                } else {
                    this.mainLoading.$emit('hide');
                }
            },
        },
        actions: {
            async submit({entityKey, endpoint, dataFormatter=noop }={}) {
                if(entityKey && entityKey !== this.entityKey || this.pendingJobs.length) return;

                try {
                    const { data } = await this.post(endpoint, dataFormatter(this));
                    if(this.independant) {
                        this.$emit('submitted', data);
                    }
                    else if(data.ok) {
                        this.mainLoading.$emit('show');
                        this.redirectToList();
                    }
                }
                catch(error) {
                    this.handleError(error);
                }
            },
            async 'delete'() {
                try {
                    await this.axiosInstance.delete(this.apiPath);
                    this.redirectToList();
                }
                catch(error) {

                }
            },
            cancel() {
                this.redirectToList();
            },
            localeChanged(newLocale) {
                this.locale = newLocale;
            },
            reset({ entityKey }={}) {
                if(entityKey && entityKey !== this.entityKey) return;

                this.data = {};
                this.errors = {};
            },

            setPendingJob({ key, origin, value:isPending }) {
                if(isPending)
                    this.pendingJobs.push(key);
                else
                    this.pendingJobs = this.pendingJobs.filter(jobKey => jobKey !== key);

                if(this.pendingJobs.length) {
                    this.actionsBus.$emit('updateActionsState', {
                        state: 'pending',
                        modifier: origin
                    })
                }
                else {
                    this.actionsBus.$emit('updateActionsState', null);
                }
            }
        },
        created() {
            this.$on('error-cleared', errorId => {
                if(this.errors[errorId])
                    this.$set(this.errors[errorId],'cleared',true);
            })
        },
        mounted() {
            this.init();
        }
    }
</script>