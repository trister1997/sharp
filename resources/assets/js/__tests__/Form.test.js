import Vue from 'vue';
import Form from '../components/form/Form.vue';
import * as fieldContainerModule from '../components/form/FieldContainer';

import * as gridModule from '../components/Grid.vue';
import * as fieldsLayoutModule from '../components/form/FieldsLayout.vue';
import * as tabbedLayoutModule from '../components/TabbedLayout.vue';


import * as consts from '../consts';


import { mockSFC, unmockSFC, wait } from "./test-utils";

import moxios from 'moxios';
import {MockInjections, MockI18n} from "./test-utils";
import { nextRequestFulfilled } from './test-utils/moxios-utils';


describe('sharp-form', ()=>{
    Vue.use(MockI18n);
    MockI18n.mockLangFunction();

    Vue.component('sharp-form', Form);
    Vue.component('sharp-field-display', mockSFC(fieldContainerModule));

    const oldDelay = moxios.delay;
    beforeAll(()=>{
        mockSFC(gridModule,{
            template: `
            <div id="MOCKED_GRID">
                <slot v-bind="rows[0][0]"></slot>
            </div>
            `
        });

        mockSFC(fieldsLayoutModule, {
            template: `
            <div id="MOCKED_FIELDS_LAYOUT">
                <slot v-if="layout[0][0]" v-bind="layout[0][0]"></slot>
            </div>
            `
        });

        mockSFC(tabbedLayoutModule, {
            template:`
            <div id="MOCKED_TABBED_LAYOUT">
                <slot v-bind="layout.tabs[0]"></slot>
            </div>
            `
        });
        moxios.delay = 10;
    });

    afterAll(()=>{
        unmockSFC(gridModule);
        unmockSFC(tabbedLayoutModule);
        unmockSFC(fieldsLayoutModule);
        unmockSFC(fieldContainerModule);
        moxios.delay = oldDelay;
    });

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-form 
                    entity-key="spaceship" 
                    :instance-id="instanceId"
                    :independant="independant" 
                    :props="props"
                    :ignore-authorizations="ignoreAuthorizations">
                </sharp-form>
            </div>
        `;
    });

    afterEach(()=>{
        moxios.uninstall();
    });

    test('can mount sharp-form', async ()=>{
        await createVm();

        await nextRequestFulfilled({
            status: 200,
            response: {
                data:{
                    title: null
                },
                fields: {
                    title: {
                        type: 'text'
                    }
                },

                layout: {
                    tabs : [
                        {
                            columns: [
                                {
                                    fields: [
                                        [{ key: 'title'}]
                                    ]
                                }
                            ]
                        }
                    ]
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "independant" sharp-form', async ()=>{
        await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },

                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount sharp-form with alert', async ()=>{
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },

                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        $form.errors = {
            title: ['required']
        };

        await Vue.nextTick();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    xtest('api path', async ()=> {
        consts.API_PATH = '/test-api';
        let $form = await createVm();

        let { $root:vm } = $form;

        expect($form.apiPath).toBe('/test-api/form/spaceship');

        vm.instanceId = '10';

        await Vue.nextTick();

        expect($form.apiPath).toBe('/test-api/form/spaceship/10');
    });

    test('localized', async ()=>{
        let $form = await createVm();
        expect($form.localized).toBe(false);
        $form.locales = ['fr', 'en'];
        await Vue.nextTick();
        expect($form.localized).toBe(true);
    });

    test('detect when is creation', async ()=>{
        let $form = await createVm();

        let { $root:vm } = $form;

        expect($form.isCreation).toBe(true);

        vm.instanceId = '10';

        await Vue.nextTick();

        expect($form.isCreation).toBe(false);

    });

    test('is read only', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    layout:{ tabs:[{ columns:[{ fields:[[]] }] }] },
                    fields:{},
                    data:{},
                    authorizations: {
                        create: false,
                        update: false
                    }
                }
            }
        });

        let { $root:vm } = $form;

        expect($form.isReadOnly).toBe(true);

        Vue.set($form.authorizations, 'create', true);

        expect($form.isReadOnly).toBe(false);

        vm.instanceId = '10';

        await Vue.nextTick();

        expect($form.isReadOnly).toBe(true);

        Vue.set($form.authorizations, 'update', true);

        expect($form.isReadOnly).toBe(false);

        vm.ignoreAuthorizations = true;
        $form.authorizations = { create:false, update:false };

        await Vue.nextTick();

        expect($form.isReadOnly).toBe(false);

    })

    test('is synchronous', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    layout:{ tabs:[{ columns:[{ fields:[[]] }] }] },
                    fields:{},
                    data:{},
                }
            }
        });

        expect($form.synchronous).toBe(true);
    });

    test('is asynchronous', async () => {
        let $form = await createVm();

        expect($form.synchronous).toBe(false);
    });


    test('has errors', async () => {
        let $form = await createVm();
        expect($form.hasErrors).toBe(false);

        $form.errors = { field: ['required'] };

        expect($form.hasErrors).toBe(true);

        Vue.set($form.errors.field, 'cleared', true);

        expect($form.hasErrors).toBe(false);
    });

    test('locale selector errors', async ()=> {
        let $form = await createVm();
        $form.locales = ['fr', 'en', 'de'];
        $form.errors = {
            'label': 'error',
            'title.fr': 'error',
        };
        expect($form.localeSelectorErrors).toEqual({ 'fr':true });

        $form.errors = {
            'label': 'error'
        };
        expect($form.localeSelectorErrors).toEqual({ });
    });

    test('expose appropriate props to layout components', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    layout:{ tabs:[{ columns:[{ fields:[[{key:'title'}]] }] }] },
                    fields:{ title: { type:'text' } },
                    data:{},
                }
            }
        });

        let { tabbedLayout, columnsGrid, fieldLayout } = $form.$refs;

        expect(tabbedLayout.$options.propsData).toEqual({
            layout: { tabs:[{ columns:[{ fields:[[{key:'title'}]] }] }] } // $form.layout
        });
        expect(columnsGrid.$options.propsData).toEqual({
            rows: [[{ fields:[[{key:'title'}]] }]] // [$form.layout.tabs[0].columns]
        });
        expect(fieldLayout.$options.propsData).toEqual({
            layout: [[{key:'title'}]], // $form.layout.tabs[0].columns[0].fields
            visible: { title: true } // $form.fieldVisible
        });
    });

    test('expose appropriate props to field', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            key: 'title',
                            type: 'text',
                            localized: true,
                        }
                    },
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    },
                    locales: ['fr', 'en']
                }
            }
        });

        let { field } = $form.$refs;

        expect(field.$options.propsData).toMatchObject({
            fieldKey: 'title',
            fieldLayout: { key: 'title' },
            updateData: $form.updateData,
            locale: 'fr'
        });

        expect(field.$attrs).toMatchObject({
            'error-identifier': 'title',
            'config-identifier': 'title',
            'update-visibility': $form.updateVisibility,
            'context-data': {
                title: null
            },
            'context-fields': {
                title: {
                    type: 'text',
                    readOnly: true
                },
            },
        });

        $form.authorizations = { create: true };

        await Vue.nextTick();

        expect(field.$attrs['context-fields'].readOnly).toBeFalsy();
    });

    test('update data', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });
        $form.fieldLocalizedValue = jest.fn(()=>'fieldLocalizedValue');
        expect($form.data.title).toBe(null);

        $form.updateData('title', 'text');
        expect($form.fieldLocalizedValue).toHaveBeenCalledWith('title', 'text');
        expect($form.data.title).toBe('fieldLocalizedValue');
    });


    test('update visibility', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        expect($form.fieldVisible.title).toBe(true);

        $form.updateVisibility('title', false);

        expect($form.fieldVisible.title).toBe(false);
    });

    test('mount', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            key: 'title',
                            type: 'text',
                            localized: true,
                        }
                    },
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ key: 'title'}]
                                        ]
                                    }
                                ]
                            }
                        ]
                    },
                    authorizations: {
                        create: true,
                        update: false
                    },
                    locales: ['en', 'fr']
                }
            }
        });

        expect($form).toMatchObject({
            data:{
                title: null
            },
            fields: {
                title: {
                    type: 'text',
                }
            },
            layout: {
                tabs : [
                    {
                        columns: [
                            {
                                fields: [
                                    [{ key: 'title' }]
                                ]
                            }
                        ]
                    }
                ]
            },
            authorizations: {
                create: true,
                update: false
            },
            locales: ['en', 'fr'],
            fieldLocale: {
                title: 'en'
            }
        });

        $form.patchLayout = ()=>{};
        $form.ready = false;
        $form.mount({ fields:{}, locales:null });

        expect($form.fieldLocale).toEqual({ title:undefined });
    });

    test('mount async', async () => {
        let $form = await createVm();

        await nextRequestFulfilled({
            status: 200,
            response: {
                data:{
                    title: null
                },
                fields: {
                    title: {
                        type: 'text'
                    }
                },
                layout: {
                    tabs : [
                        {
                            columns: [
                                {
                                    fields: [
                                        [{ key: 'title'}]
                                    ]
                                }
                            ]
                        }
                    ]
                },
                authorizations: {
                    create: true,
                    update: false
                }
            }
        });

        expect($form).toMatchObject({
            data:{
                title: null
            },
            fields: {
                title: {
                    type: 'text'
                }
            },
            layout: {
                tabs : [
                    {
                        columns: [
                            {
                                fields: [
                                    [{ key: 'title'}]
                                ]
                            }
                        ]
                    }
                ]
            },
            authorizations: {
                create: true,
                update: false
            }
        });
    });

    test('handle 422', async () => {
        let $form = await createVm();

        $form.actionsBus.$emit('submit');

        await nextRequestFulfilled({
            status: 422,
            response: {
                errors: {
                    title: ['invalid']
                }
            }
        });

        expect($form.errors).toEqual({
            title: ['invalid']
        });
    });

    test('patch layout', async () => {
        let $form = await createVm({
            propsData:{
                independant: true,
                props: {
                    data:{
                        title: null
                    },
                    fields: {
                        title: {
                            type: 'text'
                        }
                    },
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [
                                            [{ legend:'fieldset', fields: [[]] }],
                                            [{ legend:'fieldset', fields: [[]] }]
                                        ]
                                    }
                                ]
                            }
                        ]
                    }
                }
            }
        });

        let {
            tabs : [
                {
                    columns: [{ fields }]
                }
            ]
        } = $form.layout;

        expect(typeof fields[0][0].id).toBe('string');
        expect(fields[0][0].id).not.toEqual(fields[1][0].id)
    });


    test('setup action bar correctly', async () => {
        let $form = await createVm();

        let { $root:vm } = $form;

        let setupEmitted = jest.fn();
        $form.actionsBus.$on('setup', setupEmitted);
        $form.setupActionBar = jest.fn($form.setupActionBar);

        await nextRequestFulfilled({
            status: 200,
            response: {
                data:{
                    title: null
                },
                fields: {
                    title: {
                        type: 'text'
                    }
                },
                layout: {
                    tabs : [
                        {
                            columns: [
                                {
                                    fields: [
                                        [{ key: 'title'}]
                                    ]
                                }
                            ]
                        }
                    ]
                },
                authorizations: {
                    create: false,
                    update: false
                }
            }
        });

        expect($form.setupActionBar).toHaveBeenCalledTimes(1);
        expect(setupEmitted).toHaveBeenCalledTimes(1);
        expect(setupEmitted.mock.calls[0][0]).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: false,
            showBackButton: true,
            opType: 'create'
        });


        $form.authorizations.create = true;

        $form.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(2);
        expect(setupEmitted.mock.calls[1][0]).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: false,
            showBackButton: false,
            opType: 'create'
        });


        vm.instanceId = '10';
        $form.authorizations.delete = true;

        await Vue.nextTick();

        $form.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(3);
        expect(setupEmitted.mock.calls[2][0]).toMatchObject({
            showSubmitButton: false,
            showDeleteButton: true,
            showBackButton: true,
            opType: 'update'
        });

        $form.authorizations.update = true;

        $form.setupActionBar();

        expect(setupEmitted).toHaveBeenCalledTimes(4);
        expect(setupEmitted.mock.calls[3][0]).toMatchObject({
            showSubmitButton: true,
            showDeleteButton: true,
            showBackButton: false,
            opType: 'update'
        });

    });

    test('redirect to list', async () => {
        let $form = await createVm();

        expect($form.listUrl).toEqual('/sharp/list/spaceship?restore-context=1');

        $form.redirectToList = jest.fn();
        $form.actionsBus.$emit('submit');

        await nextRequestFulfilled({
            status: 200,
            response: {
                ok: true
            }
        }, 0);

        expect($form.redirectToList).toHaveBeenCalledTimes(1);

        $form.actionsBus.$emit('delete');

        await nextRequestFulfilled({
            status: 200
        }, 0);

        expect($form.redirectToList).toHaveBeenCalledTimes(2);

        $form.actionsBus.$emit('cancel');

        expect($form.redirectToList).toHaveBeenCalledTimes(3);
    });

    test('submit', async () => {
        let $form = await createVm({
            propsData: {
                independant:true,
                props: {
                    data:{},
                    fields: {},
                    layout: {
                        tabs : [
                            {
                                columns: [
                                    {
                                        fields: [[]]
                                    }
                                ]
                            }
                        ]
                    }
                }
            },
        });

        $form.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));

        $form.pendingJobs.push('upload');
        $form.actionsBus.$emit('submit');

        await wait(10);

        expect($form.post).not.toHaveBeenCalled();


        let submittedEmmitted = jest.fn();
        $form.pendingJobs = [];

        $form.actionsBus.$emit('submit');

        await wait(10);

        expect($form.post).toHaveBeenCalledTimes(1);
    });

    test('dependant submit', async () => {
        let $form = await createVm();

        $form.post = jest.fn(()=>Promise.resolve({ data: { ok: true } }));
        $form.handleError = jest.fn();

        $form.actionsBus.$emit('submit');

        await wait(10);

        expect($form.post).toHaveBeenCalledTimes(1);
        expect($form.post.mock.calls[0][0]).toBeUndefined();
        expect($form.post.mock.calls[0][1]).toBeUndefined();


        expect($form.handleError).not.toHaveBeenCalled();

        $form.post = jest.fn(()=>Promise.reject({ error: true }));

        $form.actionsBus.$emit('submit');

        await wait(10);

        expect($form.handleError).toHaveBeenCalledTimes(1);
        expect($form.handleError).toHaveBeenCalledWith({ error: true });
    });

    test('delete', async ()=>{
        // refer 'redirect to list' test
    });

    test('cancel', async ()=>{
        // refer 'redirect to list' test
    });

    test('pending jobs', async ()=> {
        let $form = await createVm();
        let updateActionsStateEmitted = jest.fn();

        $form.actionsBus.$on('updateActionsState', updateActionsStateEmitted);

        $form.actionsBus.$emit('setPendingJob', { key:'myUpload', origin:'upload' ,value:true });

        expect($form.pendingJobs).toEqual(['myUpload']);
        expect(updateActionsStateEmitted).toHaveBeenLastCalledWith({
            state: 'pending',
            modifier: 'upload'
        });

        $form.actionsBus.$emit('setPendingJob', { key:'myUpload', origin:'upload' ,value:false });

        expect($form.pendingJobs).toEqual([]);
        expect(updateActionsStateEmitted).toHaveBeenLastCalledWith(null);
    });

    test('has localize mixin with right fieldsProps', async () => {
        let $list = await createVm();
        expect($list.$options._localizedForm).toBe('fields');
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [MockInjections, customOptions],

        props: ['independant', 'props'],

        'extends': {
            data:()=>({
                instanceId: null,
                ignoreAuthorizations: null
            })
        },

        created() {
            let { axiosInstance } = this._provided;
            moxios.install(axiosInstance);
            moxios.uninstall = moxios.uninstall.bind(moxios, axiosInstance);
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}