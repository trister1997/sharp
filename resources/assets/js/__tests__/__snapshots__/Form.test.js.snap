// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sharp-form can mount "independant" sharp-form 1`] = `
<div id="app">
	<div class="SharpForm">
		<div class="container">
			<div role="alert" class="SharpNotification SharpNotification--error" style="display: none;">
				<div class="SharpNotification__details">
					<div class="SharpNotification__text-wrapper">
						<p class="SharpNotification__title">{{ form.validation_error.title }}</p>
						<p class="SharpNotification__subtitle">{{ form.validation_error.description }}</p>
					</div>
				</div>
			</div>
			<div id="MOCKED_TABBED_LAYOUT">
				<div id="MOCKED_GRID">
					<div id="MOCKED_FIELDS_LAYOUT">
						__MOCKED_SFC_SharpFieldContainer__
					</div>
				</div>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form 1`] = `
<div id="app">
	<div class="SharpForm">
		<div class="container">
			<div role="alert" class="SharpNotification SharpNotification--error" style="display: none;">
				<div class="SharpNotification__details">
					<div class="SharpNotification__text-wrapper">
						<p class="SharpNotification__title">{{ form.validation_error.title }}</p>
						<p class="SharpNotification__subtitle">{{ form.validation_error.description }}</p>
					</div>
				</div>
			</div>
			<div id="MOCKED_TABBED_LAYOUT">
				<div id="MOCKED_GRID">
					<div id="MOCKED_FIELDS_LAYOUT">
						__MOCKED_SFC_SharpFieldContainer__
					</div>
				</div>
			</div>
		</div>
	</div>
</div>
`;

exports[`sharp-form can mount sharp-form with alert 1`] = `
<div id="app">
	<div class="SharpForm">
		<div class="container">
			<div role="alert" class="SharpNotification SharpNotification--error" style="">
				<div class="SharpNotification__details">
					<div class="SharpNotification__text-wrapper">
						<p class="SharpNotification__title">{{ form.validation_error.title }}</p>
						<p class="SharpNotification__subtitle">{{ form.validation_error.description }}</p>
					</div>
				</div>
			</div>
			<div id="MOCKED_TABBED_LAYOUT">
				<div id="MOCKED_GRID">
					<div id="MOCKED_FIELDS_LAYOUT">
						__MOCKED_SFC_SharpFieldContainer__
					</div>
				</div>
			</div>
		</div>
	</div>
</div>
`;
